src/components/Panel.ts

```diff
@@ -74,6 +78,163 @@ class Panel {
+
+    const relativeIndex = (options.circular ? Math.floor(this.getPosition() / scrollAreaSize) * panelCount : 0) + this.getIndex();
+    const progress = relativeIndex - viewport.getCurrentProgress();
+
+    return progress;
+  }
+
+  public getOutsetProgress() {
+    const viewport = this.viewport;
+    const outsetRange = [
+      -this.getSize(),
+      viewport.getRelativeHangerPosition() - this.getRelativeAnchorPosition(),
+      viewport.getSize(),
+    ];
+    const relativePanelPosition = this.getPosition() - viewport.getCameraPosition();
+    const outsetProgress = getProgress(relativePanelPosition, outsetRange);
+
+    return outsetProgress;
+  }
+
+  public getVisibleRatio() {
+    const viewport = this.viewport;
+    const panelSize = this.getSize();
+    const relativePanelPosition = this.getPosition() - viewport.getCameraPosition();
+    const rightRelativePanelPosition = relativePanelPosition + panelSize;
+
+    const visibleSize = Math.min(viewport.getSize(), rightRelativePanelPosition) - Math.max(relativePanelPosition, 0);
+    const visibleRatio = visibleSize >= 0
+      ? visibleSize / panelSize
+      : 0;
+
+    return visibleRatio;
+  }
+
+  public focus(duration?: number): void {
+    const viewport = this.viewport;
+    const currentPanel = viewport.getCurrentPanel();
+    const hangerPosition = viewport.getHangerPosition();
+    const anchorPosition = this.getAnchorPosition();
+    if (hangerPosition === anchorPosition || !currentPanel) {
+      return;
+    }
+
+    const currentPosition = currentPanel.getPosition();
+    const eventType = currentPosition === this.getPosition()
+      ? ""
+      : EVENTS.CHANGE;
+
+    viewport.moveTo(this, eventType, null, duration);
+  }
+
+  public update(updateFunction: (element: HTMLElement) => any): void {
+    this.getIdenticalPanels()
+      .forEach(eachPanel => updateFunction(eachPanel.getElement()));
+  }
+
+  public prev(): FlickingPanel | null {
+    const viewport = this.viewport;
+    const options = viewport.options;
+    const prevSibling = this.prevSibling;
+
+    if (!prevSibling) {
+      return null;
+    }
+
+    const currentIndex = this.getIndex();
+    const currentPosition = this.getPosition();
+    const prevPanelIndex = prevSibling.getIndex();
+    const prevPanelPosition = prevSibling.getPosition();
+    const prevPanelSize = prevSibling.getSize();
+
+    const hasEmptyPanelBetween = currentIndex - prevPanelIndex > 1;
+    const notYetMinPanel = options.infinite
+      && currentIndex > 0
+      && prevPanelIndex > currentIndex;
+
+    if (hasEmptyPanelBetween || notYetMinPanel) {
+      // Empty panel exists between
+      return null;
+    }
+
+    const newPosition = currentPosition - prevPanelSize - options.gap;
+
+    let prevPanel = prevSibling;
+    if (prevPanelPosition >= currentPosition) {
+      prevPanel = prevSibling.clone(prevSibling.getCloneIndex(), true);
+      prevPanel.setPosition(newPosition, true);
+    }
+
+    return prevPanel;
```

Comment:
@WoodNeck 
It's a special case.
In circular mode, if enough virtual clones are created through prev() calls, then next(), the clones are not created.

Reply:
Okay, I forgot to consider that one. I'll push fixing commit for it.

Reply:
I couldn't reproduce condition you mentioned. Can you give me some more details?

Reply:
Okay, now I got what's the problem.
I'll push fixing commit for it.